src/distdl/nn/conv_feature_new.py

```diff
@@ -0,0 +1,388 @@
+            del self.conv_layer.bias
+            self.conv_layer.bias = new_bias
+
+        self.w_broadcast = Broadcast(self.P_wb_cart, self.P_x,
+                                     preserve_batch=False)
+
+        if self.conv_layer.bias is not None:
+            self.b_broadcast = Broadcast(self.P_wb_cart, self.P_x,
+                                         preserve_batch=False)
+
+        # We need to pad the input ourselves, since ConvNd can only pad symmetrically.
+        # Since global dimensions are needed, construction defered to pre-forward hook.
+        self.pad_layer = None
+
+        # We need the halo shape, and other info, to fully populate the pad,
+        # halo exchange, and unpad layers.  For pad and unpad, we defer their
+        # construction to the pre-forward hook.
+        self.torch_halo_pad = None
+
+        # We need to be able to remove some data from the input to the conv
+        # layer.
+        self.needed_slices = None
+
+        # For the halo layer we also defer construction, so that we can have
+        # the halo shape for the input.  The halo will allocate its own
+        # buffers, but it needs this information at construction to be able
+        # to do this in the pre-forward hook.
+        self.halo_layer = None
+
+        # Variables for tracking input changes and buffer construction
+        self._distdl_is_setup = False
+        self._input_tensor_structure = TensorStructure()
+
+    def _distdl_module_setup(self, input):
+        r"""Distributed (feature) convolution module setup function.
+
+        This function is called every time something changes in the input
+        tensor structure.  It should not be called manually.
+
+        Parameters
+        ----------
+        input :
+            Tuple of forward inputs.  See
+            `torch.nn.Module.register_forward_pre_hook` for more details.
+
+        """
+
+        self._distdl_is_setup = True
+        self._input_tensor_structure = TensorStructure(input[0])
+
+        if not self.P_x.active:
+            return
+
+        if self.serial:
+            return
+
+        dims = len(self.padding)
+        pad_left_right = self.padding.reshape((dims, 1)) + np.zeros((dims, 2), dtype=np.int)
+        self.pad_layer = DistributedPad(self.P_x, pad_left_right, mode='constant', value=0)
+
+        # Compute the global and local input shapes after padding
+        x_global_shape = self._distdl_backend.assemble_global_tensor_structure(input[0], self.P_x).shape
+        x_local_shape = np.array(input[0].shape)
+        x_global_shape_after_pad = x_global_shape + 2*self.padding
+        x_local_shape_after_pad = x_local_shape + np.sum(self.pad_layer.local_pad, axis=1, keepdims=False)
+
+        # Compute left and right start and stop indices for the local input shape
+        subtensor_shapes_unbalanced = \
+            compute_subtensor_shapes_unbalanced(TensorStructure(shape=x_local_shape_after_pad), self.P_x)
```

Comment:
Here, you've already done the communication necessary to figure out the neighbor halo sizes.